### PR TITLE
Clean up and refactor help generator

### PR DESCRIPTION
**Pull request by jougs**
This PR refactors the help generator to be more generic and also usable for extension modules.
- Install the help generator
- Add documentation extraction to the MyModule example
- Use commandline arguments to pass absolute paths to the parser instead of hardcoding relative ones
- Put generated help files in the build directory instead of  the source directory

This PR is to be applied after #2 is merged.

There's one problem left: the generation of the help index files currently only has the functions from the last run of the parser. This needs to be changed so that all files present in the installation's help directory are included in the index files.

### CODE REVIEW

**Comment by apeyser** [comment]
@jougs Why is this so big? Is this a rebase of another set of patches?

**Comment by jougs** [comment]
I've only merged in master. Apparently that was not done for quite some time on the base branch. I think the unrelated commits will disappear from this PR if @steffengraber did that. I've added a corresponding comment to https://github.com/nest/nest-simulator/pull/464.

File: extras/help_generator/helpers.py [hunk 1]
```diff
@@ -51,3 +53,29 @@ def check_ifdef(item, filetext, docstring):
...
+    Forgiving version of os.makedirs, emulating the behavior of the
+    shell command 'mkdir -p'. The function tries to create the
+    directory at the given path including all subdirectories and
+    returns silently if the directory already exists.
+    """
+
+    try:
+        os.makedirs(path)
+    except OSError as exc:
+        if exc.errno == errno.EEXIST and os.path.isdir(path):
```
Comment (apeyser) [inline extras/help_generator/helpers.py:69]:
```
# Ignore exception if the directory already exists
if exc.errno != errno.EEXIST or os.path.isdir(path):
    raise
```

File: extras/help_generator/parse_help.py [hunk 2]
```diff
@@ -62,14 +71,13 @@
 # Now begin to collect the data for the helpindex.html
 
 dcs = r'\/\*[\s?]*[\n?]*BeginDocumentation[\s?]*\:?[\s?]*[.?]*\n(.*?)\n*?\*\/'
-for file in allfiles:
-    if not file.endswith('.py'):
-        f = open(('%s' % (file,)), 'r')
+for fname in allfiles:
+    if not fname.endswith('.py'):
+        f = open(('%s' % (fname,)), 'r')
```
Comment (apeyser) [inline extras/help_generator/parse_help.py:76]: Why the complexity here? Isn't this just `open(fname, 'r')`?

File: extras/help_generator/parse_help.py [hunk 3]
```diff
@@ -81,32 +89,33 @@
...
 
 # Now begin to collect the data for the help files and start generating.
 dcs = r'\/\*[\s?]*[\n?]*BeginDocumentation[\s?]*\:?[\s?]*[.?]*\n(.*?)\n*?\*\/'
-for file in allfiles:
+for fname in allfiles:
     # .py is for future use
-    if not file.endswith('.py'):
-        f = open(('%s' % (file,)), 'r')
+    if not fname.endswith('.py'):
+        f = open(('%s' % (fname,)), 'r')
```
Comment (apeyser) [inline extras/help_generator/parse_help.py:115]: And same here: `open(fname, 'r')`

File: extras/help_generator/writers.py [hunk 4]
```diff
@@ -148,26 +135,21 @@ def write_help_html(doc_dic, file, sli_command_list, keywords):
...
-            f_file_name_hlp.close()
+        if fname.endswith('.sli'):
+            path = os.path.join(helpdir, 'sli')
         else:
-            f_file_name = open(('../cmds/cc/%s.html' % name), 'w')
-            f_file_name.write(cmdindexstring)
-            f_file_name.close()
+            path = os.path.join(helpdir, 'cc')
+
+        f_file_name = open((path + '/%s.html' % name), 'w')
```
Comment (apeyser) [inline extras/help_generator/writers.py:143]: Isn't this deprecated? Shouldn't we use `"{}/{}.html".format(path, name)`?

File: extras/help_generator/writers.py [hunk 5]
```diff
@@ -148,26 +135,21 @@ def write_help_html(doc_dic, file, sli_command_list, keywords):
...
+        f_file_name.write(cmdindexstring)
+        f_file_name.close()
 
-            f_file_name_hlp = open(('../cmds/cc/%s.hlp' % name), 'w')
-            f_file_name_hlp.write('\n'.join(hlplist))
-            f_file_name_hlp.close()
-        # return name
+        f_file_name_hlp = open((path + '/%s.hlp' % name), 'w')
```
Comment (apeyser) [inline extras/help_generator/writers.py:147]: And here: Isn't this deprecated? Shouldn't we use `"{}/{}.html".format(path, name)`?

**Pull request closed without merging**